### PR TITLE
Refactor data loading workflow and support single evaluation mode

### PR DESCRIPTION
**Pull request by sian-chen**
1. Change the data format
2. Remove code for handling MIMIC
3. Refactor data loading workflow to support single evaluation mode

### CODE REVIEW

File: data_utils.py [hunk 1]
```diff
@@ -63,118 +62,74 @@ def get_dataset_loader(config, dataset, shuffle=False, train=True):
...
+    datasets['train'] = _load_raw_data(train_path)
+    val_path = config[f'val_path'] or os.path.join(config.data_dir, 'valid.txt')
+    if os.path.exists(val_path):
+        datasets['val'] = _load_raw_data(val_path)
+    else:
+        datasets['train'], datasets['val'] = train_test_split(
+            datasets['train'], test_size=config.val_size, random_state=42)
 
-    log.info(f"Finish loading dataset (train: {len(datasets['train'])} / test: {len(datasets['test'])} / dev: {len(datasets['dev'])})")
+    log.info(f"Finish loading dataset (train: {len(datasets['train'])} / test: {len(datasets['test'])} / val: {len(datasets['val'])})")
```
Comment (Eleven1Liu) [inline data_utils.py:98]: What if the user does not specify a `test.txt` in the training mode?

**Review by Eleven1Liu — CHANGES_REQUESTED** [review]

File: data_utils.py (hunk 1 above)
Comment (sian-chen) [reply data_utils.py:98]: The program will do nothing after training.
Please check L138 in main.py.

Comment (Eleven1Liu) [reply data_utils.py:98]: Yes, but in this `log.info` there will not have`test` key in the `datasets` which give a `KeyError`?

File: README.md [hunk 2]
```diff
@@ -105,9 +101,15 @@ To use your own word embeddings or vocabulary set, specify the following paramet
...
+In the validation progress, you can evaluate the model with a set of evaluation metrics. Set `monitor_metrics` to define what you want to print on the screen. `val_metric` is the metric for picking the best model. Example:
 ```yaml
 monitor_metrics: [P@1, P@3, P@5]
 val_metric: P@1
 ```
+
+### Evaluation
+In the evaluation progress, you can evaluate a model from a pre-obtained checkpoint.
+```
+python main.py --eval --load_checkpoint CHECKPOINT_PATH --test_path TEST_DATA_PATH
```
Comment (Eleven1Liu) [inline README.md:114]:
```suggestion
python main.py --config CONFIG_PATH --eval --load_checkpoint CHECKPOINT_PATH --test_path TEST_DATA_PATH
```

Comment (sian-chen) [reply README.md:114]: We don't need a config file in evaluation mode now.

Comment (Eleven1Liu) [reply README.md:114]: Or should we add parameters like `--filter_size` to avoid size mismatch when loading `state_dict`?

File: README.md [hunk 3]
```diff
@@ -65,26 +65,22 @@ Each parameter can also be specified through the command line.
 
 ## Data Format
 
-Put texts and labels in the training and test set separately in `train_text.txt`, `test_text.txt`, `train_labels.txt`, and `test_labels.txt`. The program internally splits the training set to two parts for training and validation. To specify the size of the validation set, use `dev_size` in one of the following two ways:
+Put texts and labels in the training, validation, and test set separately in `train.txt`, `valid.txt`, and `test.txt`, or specifying the path by the arguments  `train_path`, `valid_path`, and `test_path`. The program internally splits the training set to two parts for training and validation. To specify the size of the validation set, use `dev_size` in one of the following two ways:
```
Comment (cjlin1) [inline README.md:68]: If validation set is not provided, then the program internally splits ...

File: README.md [hunk 4]
```diff
@@ -65,26 +65,22 @@ Each parameter can also be specified through the command line.
...
 - a ratio in [0,1] to indicate the percentage of training data allocated for validation
 - an integer to indicate the number of training data allocated for validation
 
-### Examples of a text file and a label file: 
-- **texts**: one sample per line
+### Examples of a training file:
+- one sample per line
+- seperate labels and texts by `<TAB>`
+- labels are split by spaces
```
Comment (cjlin1) [inline README.md:75]: but we also want to support ID<tab>labels<tab>texts??

File: README.md [hunk 5]
```diff
@@ -65,26 +65,22 @@ Each parameter can also be specified through the command line.
...
-recov recov recov recov excit ...
-uruguay uruguay compan compan compan ...
-```
-
-- **labels**: one sample per line; labels are split by space
-```
-E11 ECAT M11 M12 MCAT
-C24 CCAT
+E11 ECAT M11 M12 MCAT<TAB>recov recov recov recov excit ...
+C24 CCAT<TAB>uruguay uruguay compan compan compan ...
```
Comment (cjlin1) [inline README.md:78]: Need a description about labels in test files. You can modify things from libsvm description "In the test set, <label> is used only to calculate accuracy or errors. If it's unknown, any number is fine."

File: data_utils.py (hunk 1 above)
Comment (sian-chen) [reply data_utils.py:98]: Oops, you're right!

File: README.md (hunk 2 above)
Comment (sian-chen) [reply README.md:114]: Nice catch. Specifying a config file would be easier for now.

File: README.md (hunk 3 above)
Comment (sian-chen) [reply README.md]: Fixed.

File: README.md (hunk 4 above)
Comment (sian-chen) [reply README.md:75]: Yes, the README is updated.

File: README.md (hunk 5 above)
Comment (sian-chen) [reply README.md:86]: Fixed.

File: data_utils.py [hunk 6]
```diff
@@ -63,118 +62,76 @@ def get_dataset_loader(config, dataset, shuffle=False, train=True):
...
 
-    for split in datasets.keys():
-        datasets[split] = TextDataset(datasets[split], text_dict, classes, config['max_seq_length'])
+@log.enter('load_dataset')
+def load_datasets(config):
+    datasets = {}
+    test_path = config.test_path or os.path.join(config.data_dir, 'test.txt')
+    if os.path.exists(test_path):
+        datasets['test'] = _load_raw_data(test_path, is_test=True)
+    if config.eval:
```
Comment (cjlin1) [inline data_utils.py:87]: what if config.eval is true but test_path does not exist? Need an error message

Comment (sian-chen) [reply data_utils.py:87]: Fixed.

File: data_utils.py [hunk 7]
```diff
@@ -63,118 +62,76 @@ def get_dataset_loader(config, dataset, shuffle=False, train=True):
...
+        return datasets
```
Comment (cjlin1) [inline data_utils.py:88]: Using return here causes that msg not printed in line 99-100?
Maybe a flow as follows?

if eval
   if test exists ...
   else error
else
   if test exists ...
   load training

So we only return in the end?

File: data_utils.py [hunk 8]
```diff
@@ -63,118 +62,76 @@ def get_dataset_loader(config, dataset, shuffle=False, train=True):
...
-        datasets['train'], datasets['dev'] = torch.utils.data.random_split(datasets['train'], [train_size, dev_size], generator=torch.Generator().manual_seed(42))
 
-    for split in datasets.keys():
-        datasets[split] = TextDataset(datasets[split], text_dict, classes, config['max_seq_length'])
+@log.enter('load_dataset')
+def load_datasets(config):
+    datasets = {}
+    test_path = config.test_path or os.path.join(config.data_dir, 'test.txt')
+    if config.eval or os.path.exists(test_path):
+        datasets['test'] = _load_raw_data(test_path, is_test=True)
```
Comment (cjlin1) [inline data_utils.py:86]: So we rely one _load_raw_data to give an error if eval=true and test_path does not exist??  What kind of error message is produced from _load_raw_data?

File: README.md [hunk 9]
```diff
@@ -65,26 +65,31 @@ Each parameter can also be specified through the command line.
 
 ## Data Format
 
-Put texts and labels in the training and test set separately in `train_text.txt`, `test_text.txt`, `train_labels.txt`, and `test_labels.txt`. The program internally splits the training set to two parts for training and validation. To specify the size of the validation set, use `dev_size` in one of the following two ways:
+Put texts and labels in the training, validation, and test set separately in `train.txt`, `valid.txt`, and `test.txt`, or specifying the path by the arguments  `train_path`, `valid_path`, and `test_path`. If validation set is not provided, then the program internally splits the training set to two parts for training and validation. To specify the size of the validation set, use `dev_size` in one of the following two ways:
```
Comment (cjlin1) [inline README.md:68]: use 'val_size'

File: README.md [hunk 10]
```diff
@@ -65,26 +65,31 @@ Each parameter can also be specified through the command line.
...
 
-- **labels**: one sample per line; labels are split by space
+Without ID column:
 ```
-E11 ECAT M11 M12 MCAT
-C24 CCAT
+E11 ECAT M11 M12 MCAT<TAB>recov recov recov recov excit ...
+C24 CCAT<TAB>uruguay uruguay compan compan compan ...
 ```
+In the test set, the labels are used to calculate accuracy or errors. If it's unknown, an any string (or empty) is fine.
```
Comment (cjlin1) [inline README.md:88]: For the "empty" situation it seems we mean

<tab>text

or

id<tab><tab>text

If so better to give examples. User would think "empty" means

text

and they won't add <tab> in the beginning

File: data_utils.py [hunk 11]
```diff
@@ -63,118 +62,76 @@ def get_dataset_loader(config, dataset, shuffle=False, train=True):
...
+                       min_freq=1, specials_first=False)
     else:
         counter = collections.Counter()
-        for example in examples:
-            unique_tokens = set(example['text'])
-            counter.update(unique_tokens)
-        vocabs = Vocab(counter, specials=['<pad>', '<unk>'], min_freq=min_vocab_freq)
-
+        for data in dataset:
+            counter.update(set(data['text']))
```
Comment (cjlin1) [inline data_utils.py:114]: The original way of having "unique_tokens" may give better readability?

File: data_utils.py [hunk 12]
```diff
@@ -63,118 +62,76 @@ def get_dataset_loader(config, dataset, shuffle=False, train=True):
...
+                       min_freq=config.min_vocab_freq)
     log.info(f'Read {len(vocabs)} vocabularies.')
     return vocabs
 
 
+def load_or_build_label(config, datasets):
+    if config.label_file:
+        log.info('Load labels from {config.label_file}')
+        with open(config.label_file, 'r') as fp:
+            classes = sorted([s.strip() for s in fp.readlines()])
```
Comment (cjlin1) [inline data_utils.py:125]: Now we assume labels are in the data files.. So do we still need this if config.label_file?

File: evaluate.py [hunk 13]
```diff
@@ -39,31 +39,31 @@ def evaluate(config, model, dataset_loader, eval_metric, split='dev', dump=True)
...
-        train_labels = np.hstack([instance[target_label]
-                                  for instance in dataset['train']])
+        # test_labels = np.hstack([instance[target_label]
+                                 # for instance in dataset['test']])
+        # train_labels = np.hstack([instance[target_label]
+                                  # for instance in dataset['train']])
 
         self.config = config
-        self.class_num = dataset['train'].num_class
+        self.class_num = config.num_class
```
Comment (cjlin1) [inline evaluate.py:51]: class_num vs.  num_class not good
how about num_classes in all places?

**Review by cjlin1 — DISMISSED** [review]

File: README.md (hunk 9 above)
Comment (cjlin1) [inline README.md:68]: Put texts and labels -> Put labels and texts

as in the file labels come first

File: data_utils.py (hunk 8 above)
Comment (sian-chen) [reply data_utils.py:86]: Yes, it raises "FileNotFoundError: [Errno 2] No such file or directory: 'XXX'".
I think the information is clear enough so we don't need to raise the exception ourselves.

File: data_utils.py (hunk 12 above)
Comment (sian-chen) [reply data_utils.py:125]: It's for the zero-shot setting in the future.

File: data_utils.py (hunk 7 above)
Comment (sian-chen) [reply data_utils.py]: Solved.

File: README.md (hunk 9 above)
Comment (sian-chen) [reply README.md]: Fixed.

File: README.md (hunk 10 above)
Comment (sian-chen) [reply README.md]: Added examples.

File: README.md (hunk 9 above)
Comment (sian-chen) [reply README.md]: Solved.

File: evaluate.py (hunk 13 above)
Comment (sian-chen) [reply evaluate.py]: Solved. Eleven is working on a refactoring of evaluate.py.

**Review by Eleven1Liu — APPROVED** [review]

**Pull request merged**